capsules/src/ble_advertising_driver.rs

```diff
@@ -630,75 +629,95 @@ where
                         app.process_status = Some(BLEState::ScanningIdle);
                         app.set_next_alarm::<A::Frequency>(self.alarm.now().into_u32());
                         self.reset_active_alarm();
-                        ReturnCode::SUCCESS
+                        CommandResult::success()
                     } else {
-                        ReturnCode::EBUSY
+                        CommandResult::failure(ErrorCode::BUSY)
                     }
                 })
                 .unwrap_or_else(|err| err.into()),
 
-            _ => ReturnCode::ENOSUPPORT,
+            _ => CommandResult::failure(ErrorCode::NOSUPPORT),
         }
+        .into()
     }
 
-    fn allow_readwrite(
+    fn allow_readonly(
         &self,
         appid: kernel::AppId,
         allow_num: usize,
-        slice: Option<AppSlice<SharedReadWrite, u8>>,
-    ) -> ReturnCode {
-        match allow_num {
+        mut slice: ReadOnlyAppSlice,
+    ) -> Result<ReadOnlyAppSlice, (ReadOnlyAppSlice, ErrorCode)> {
+        let res = match allow_num {
             // Advertisement buffer
             0 => self
                 .app
                 .enter(appid, |app, _| {
-                    app.adv_data = slice;
-                    if let ReturnCode::SUCCESS = app.generate_random_address(appid) {
+                    app.generate_random_address(appid).map(|_| {
                         app.process_status = Some(BLEState::Initialized);
-                        ReturnCode::SUCCESS
-                    } else {
-                        ReturnCode::FAIL
-                    }
+                        mem::swap(&mut app.adv_data, &mut slice);
+                    })
                 })
-                .unwrap_or_else(|err| err.into()),
+                .unwrap_or_else(|err| Err(err.into())),
+
+            // Operation not supported
+            _ => Err(ErrorCode::NOSUPPORT),
+        };
 
+        match res {
+            Ok(()) => Ok(slice),
+            Err(e) => Err((slice, e)),
+        }
+    }
+
+    fn allow_readwrite(
+        &self,
+        appid: kernel::AppId,
+        allow_num: usize,
+        mut slice: ReadWriteAppSlice,
+    ) -> Result<ReadWriteAppSlice, (ReadWriteAppSlice, ErrorCode)> {
+        let res = match allow_num {
             // Passive scanning buffer
             1 => self
```

Comment:
We might want to change the index of the `allow_readwrite` buffers to start at `0` as well.

Reply:
Agree, good catch